rfd/0001-block-registry.md

```diff
@@ -0,0 +1,95 @@
+
+## Why
+
+To experiment with various storage solutions the workspace storage layer needs to be fully decoupled from the workspace runtime and a well defined interface between them has to be used. Workspaces are currently defined as docker images so the storage layer should provide container like primitives while at the same time being compatible with VMs.
+
+## Details
+
+### Requirements
+
+1. Minimal assumptions about the host OS (rules out ZFS and custom kernel modules)
+1. Layered storage - Supports layers which can be overlayed on top of each other (rules out raw files)
+1. Dozens of layers can be combined together without performance degradation (rules out qcow2 due to performance issues with too many backing files)
+1. Allows to keep its data in sparse files, no need to reformat or dedicate a disk for it (rules out SPDK blobfs/mayastor/ceph/etc...)
+1. The resulting storage needs to natively support Docker on top of it after passing it to an VM (rules out overlayfs over virtio-fs)
+1. It should both support providing storage for a VM and a Container
+
+### OverlayBD storage
+
+OverlayBD is an OCI compatible container format from Alibaba which is compatible with our requirements. This RFD will talk about using OverlayBD for container storage. The idea was validated in [this PoC](https://github.com/hocus-dev/hocus/pull/59). Hocus will store vm storage using overlaybd which will be managed by the block registry. The interface between the block registry and the workspace runtime are block devices - the registry is responsible for managing those block devices and the storage.
+
+### TCMU
+
+Overlaybd uses TCMU under the hood. We think this it's a good choice for Hocus. In the future we may easily build a custom storage format using TCMU and the rest of our code will be compatible. All block devices created by the block registry using TCMU MUST have an SCSI Serial starting with `hocusbd-`. For example `hocusbd-u2fpCGJ7`. The WWN of the TCM loop device for the block registry MUST be `naa.726163636f6f6e73`. All TCMU storage objects will be exposed under a single SAS HBA and will be forwarded to a single TCM loop target under different LUN's. OverlayBD will be forked so it will be possible to change the name of the TCMU service and UIO path in runtime so multiple block registries can run on the same host at the same time (important for CI).
+
+### Exposing TCMU to the agent container
+
+The Hocus install script NEEDS to install the following udev rule on the host OS:
+
+```
+# Hardlink hocus block devices using their serial number
+SUBSYSTEM=="block", ACTION=="add|change", ENV{ID_SCSI_SERIAL}=="hocusbd-[0-9A-Za-z]*", RUN+="/bin/sh -c 'mkdir -p /dev/hocus && ln $env{DEVNAME} /dev/hocus/$env{ID_SCSI_SERIAL}'"
+SUBSYSTEM=="block", ACTION=="remove", ENV{ID_SCSI_SERIAL}=="hocusbd-[0-9A-Za-z]*", RUN+="/bin/sh -c 'unlink /dev/hocus/$env{ID_SCSI_SERIAL}'"
```

Comment:
also rules for UIO are required

Reply:
also tcmlib has hardcoded paths to /dev/uio* -> will need to change them to /dev/hocus/uio*